text/0000-engines-component-dependencies.md

```diff
@@ -0,0 +1,97 @@
+
+# Detailed design
+
+As it is already possible to declare services as dependencies of engines, the underlying groundwork should already be in place to also allow components to be declared as dependencies. Declaring a component dependency would look and work exactly the same way as declaring service dependencies.
+
+Consider the following, using the `ember-blog` engine as an example:
+
+### Declaring Dependencies (Engine)
+
+As with services, the engine would be required to declare any components that they expected to be supplied as dependencies.
+
+```js
+// ...
+
+export default Engine.extend({
+  modulePrefix: 'ember-blog',
+  Resolver,
+
+  dependencies: {
+    components: [
+      'custom-button',
+      'custom-heading'
+    ]
+  }
+});
+```
+
+
+### Fulfilling Dependencies (Consuming Application/Engine)
+
+Consuming applications/engines would be required to fulfil the component dependencies declared by the engine. This would also work exactly the same as fulfilling service dependencies in engines today.
+
+In order to mitigate the risk of naming collisions, it would be possible to re-map component dependencies to a custom key. See the 'heading' component dependency in the example below.
+
+```js
+
+// ...
+
+App = Ember.Application.extend({
+  modulePrefix: config.modulePrefix,
+  podModulePrefix: config.podModulePrefix,
+  Resolver,
+
+  engines: {
+    emberBlog: {
+      dependencies: {
+        components: [
+          'custom-button',
+          {'heading': 'custom-heading'}
+        ]
+      }
+    }
+  }
+});
+
+// ...
+```
+
+# How We Teach This
+
+This RFC doesn't introduce any new patterns or terminology, but instead builds upon functionality that currently exists in engines today. As a result, the learning curve of either engines or Ember itself shouldn't be affected and any documentation could be added along with the engines guides themselves.
+
+# Drawbacks
+
+### Versioning / Breaking Changes
+The main drawback to this approach is the potential for breaking changes. Any changes made to dependent components in the parent application/engine would automatically flow through to the mounted engines. As a result, if a breaking change was made, it would cause issues in any mounted engines that were using the changed component.
+
+While this is a valid issue, I believe that the risk is low due to the fact that the main use case for this feature will be in distributed applications, where engines are being used as a means of allowing developers to extend the main application. These applications in a majority of cases would already be using versioning and changelogs at a higher level as a way of communicating breaking changes to their users/developers.
```

Comment:
I feel like the fact that this is for distributed applications makes the risk greater since it is difficult to ensure all parts of the application have notification of the breakage.

In a smaller, less distributed system you have control over everything and can ensure compatibility.

That said, I believe this issue also exists for the Services dependencies we currently have.